docs/src/user_docs/assembly/execution_contexts.md

```diff
@@ -0,0 +1,105 @@
+
+### Kernels
+A *kernel* defines a set of procedures which can be invoked from user contexts to be executed in the root context. Miden assembly programs are always compiled against some kernel. The default kernel is empty - i.e., it does not contain any procedures. To compile a program against a non-empty kernel, the kernel needs to be specified when instantiating the [Miden Assembler](https://crates.io/crates/miden-assembly).
+
+A kernel can be defined similarly to a regular [library module](./code_organization.md#library-modules) - i.e., it can have internal and exported procedures. However, there are some small differences between what procedures can do in a kernel module vs. what they can do in a regular library module. Specifically:
+
+- Procedures in a kernel module cannot use `call` or `syscall` instructions. This means that creating a new context from within a syscall is not possible.
+- Unlike procedures in regular library modules, procedures in a kernel module can use `caller` instruction. This instruction puts the hash of the procedure which initiated the parent context onto the stack.
+
+### Memory layout
+As mentioned earlier, procedures executed within a given context can access memory only of that context. This is true for both memory reads and memory writes.
+
+Address space of every context is the same: the smallest accessible address is $0$ and the largest accessible address is $2^{32} - 1$. Any code executed in a given context has access to its entire address space. However, by convention, we assign different meanings to different regions of the address space.
+
+For user contexts we have the following:
+- The first $2^{30}$ words (each word is 4 field elements) are assumed to be global memory.
+- The next $2^{30}$ words are reserved for memory locals of procedures executed in the same context (i.e., via the `exec` instruction).
+- The remaining address space has no special meaning.
+
+![user memory layout](../../assets/user_docs//assembly/execution_contexts/user_mem_layout.png)
+
+For the root context we have the following:
+- The first $2^{30}$ words are assumed to be global memory.
+- The next $2^{30}$ words are reserved for memory locals of procedures executed in the root context.
+- The next $2^{30}$ words are reserved for memory locals of procedures executed from within a `syscall`.
+- The remaining address space has no special meaning.
+
+![root memory layout](../../assets/user_docs//assembly/execution_contexts/root_mem_layout.png)
+
+For both types of contexts, writing directly into regions of memory reserved for procedure locals is not advisable. Instead, `loc_load`, `loc_store` and other similar dedicated instructions should be used to access procedure locals.
+
+### Example
+To better illustrate what happens as we execute procedures in different contexts, let's go over the following example.
+
+```
+kernel
+--------------------
+export.baz.2
```

Comment:
Maybe find a better name than `baz` and `bar`?

Reply:
These are classic :)